File: 3563.lexicographically-smallest-string-after-adjacent-removals.java
#
# @lc app=leetcode id=3563 lang=java
#
# [3563] Lexicographically Smallest String After Adjacent Removals
#
# @lc code=start
import java.util.*;

class Solution {
    public String lexicographicallySmallestString(String s) {
        Map<String, String> memo = new HashMap<>();
        return dfs(s, memo);
    }
    
    private String dfs(String s, Map<String, String> memo) {
        if (s.length() <= 1) return s;
        if (memo.containsKey(s)) return memo.get(s);
        String minStr = s;
        for (int i = 0; i < s.length() - 1; ++i) {
            char a = s.charAt(i), b = s.charAt(i + 1);
            if (isConsecutive(a, b)) {
                String next = s.substring(0, i) + s.substring(i + 2);
                String candidate = dfs(next, memo);
                if (candidate.compareTo(minStr) < 0) {
                    minStr = candidate;
                }
            }
        }
        memo.put(s, minStr);
        return minStr;
    }
    
    private boolean isConsecutive(char a, char b) {
        return Math.abs(a - b) == 1 || (a == 'a' && b == 'z') || (a == 'z' && b == 'a');
    }
}
# @lc code=end

File: 3563.lexicographically-smallest-string-after-adjacent-removals.go
#
# @lc app=leetcode id=3563 lang=golang
#
# [3563] Lexicographically Smallest String After Adjacent Removals
#

# @lc code=start
func lexicographicallySmallestString(s string) string {
    memo := make(map[string]string)
    var dfs func(string) string
    dfs = func(str string) string {
        if val, ok := memo[str]; ok {
            return val
        }
        res := str // Always consider the untouched string as a candidate
        n := len(str)
        for i := 0; i < n-1; i++ {
            a, b := str[i], str[i+1]
            // Check if a and b are consecutive (with wraparound)
            if (a == 'a' && b == 'z') || (a == 'z' && b == 'a') || (a-b == 1) || (b-a == 1) {
                next := str[:i] + str[i+2:]
                // Memoization ensures we do not revisit same state, preventing cycles
                sub := dfs(next)
                if sub < res {
                    res = sub
                }
            }
        }
        memo[str] = res
        return res
    }
    result := dfs(s)
    // (For general correctness, add test assertions here with sample inputs if needed)
    return result
}
# @lc code=end